File: Baitoan3/b/odd_parts_vs_self_conjugate_partitions.py
import sys
import os

def count_odd_distinct_partitions(n):
    """
    Count the number of partitions of n into an odd number of distinct parts.
    """
    # dp[i][j]: number of partitions of i into j distinct parts
    dp = [[0] * (n + 1) for _ in range(n + 1)]
    dp[0][0] = 1
    for part in range(1, n + 1):
        for total in range(n, part - 1, -1):
            for cnt in range(1, n + 1):
                dp[total][cnt] += dp[total - part][cnt - 1]
    result = 0
    for cnt in range(1, n + 1, 2):
        result += dp[n][cnt]
    return result

def count_distinct_odd_partitions(n):
    """
    Count the number of partitions of n into distinct odd parts (self-conjugate partitions).
    """
    # dp[i][j]: number of partitions of i into distinct odd parts, max part <= j
    dp = [[0] * (n + 1) for _ in range(n + 1)]
    dp[0][0] = 1
    for max_part in range(1, n + 1, 2):
        if max_part <= n:
            dp[max_part][max_part] = 1
    for total in range(1, n + 1):
        for max_part in range(1, n + 1, 2):
            if total - max_part >= 0:
                for prev in range(max_part - 2, -1, -2):
                    dp[total][max_part] += dp[total - max_part][prev]
    result = 0
    for max_part in range(1, n + 1, 2):
        result += dp[n][max_part]
    return result

def main():
    # Read n from input.txt
    try:
        with open("input.txt", "r", encoding="utf-8") as f:
            n = int(f.read().strip())
    except Exception as e:
        print("Error reading input.txt:", e)
        sys.exit(1)
    if n < 0:
        print("n must be a non-negative integer.")
        return
    odd_distinct = count_odd_distinct_partitions(n)
    self_conjugate = count_distinct_odd_partitions(n)
    print(f"Number of partitions of {n} into an odd number of distinct parts: {odd_distinct}")
    print(f"Number of self-conjugate partitions of {n} (partitions into distinct odd parts): {self_conjugate}")
    if odd_distinct == self_conjugate:
        print("They are equal.")
    else:
        print("They are different.")

if __name__ == "__main__":
    main() 

File: Project1/Baitoan3/b/odd_parts_vs_self_conjugate_partitions.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_map>
#include <cmath>
#include <algorithm>
#include <tuple>

namespace core {
    /**
     * @brief Check if n is a generalized pentagonal number and return k if so
     * @param n The number to check
     * @return std::pair<bool, int> (is_pentagonal, k_val)
     */
    std::pair<bool, int> isGeneralizedPentagonal(int n) {
        if (n <= 0) return {false, 0};
        int val = static_cast<int>(std::sqrt(24 * n + 1));
        if (val * val != 24 * n + 1) return {false, 0};
        if ((val + 1) % 6 == 0) return {true, (val + 1) / 6};
        if ((val - 1) % 6 == 0) return {true, (val - 1) / 6};
        return {false, 0};
    }
    /**
     * @brief Compute the number of partitions of n into distinct parts (p_d(n))
     * @param n The integer to partition
     * @return The number of partitions into distinct parts
     */
    int countDistinctPartitions(int n) {
        std::vector<int> dp(n + 1, 0);
        dp[0] = 1;
        for (int i = 1; i <= n; ++i) {
            for (int j = n; j >= i; --j) {
                dp[j] += dp[j - i];
            }
        }
        return dp[n];
    }
    /**
     * @brief Compute the number of partitions of n into an odd number of distinct parts (|O(n)|)
     * @param n The integer to partition
     * @return The number of partitions into an odd number of distinct parts
     */
    int countOddDistinctPartitions(int n) {
        // dp[i][j]: number of partitions of i into j distinct parts
        std::vector<std::vector<int>> dp(n + 1, std::vector<int>(n + 1, 0));
        dp[0][0] = 1;
        for (int part = 1; part <= n; ++part) {
            for (int sum = n; sum >= part; --sum) {
                for (int cnt = 1; cnt <= n; ++cnt) {
                    dp[sum][cnt] += dp[sum - part][cnt - 1];
                }
            }
        }
        int result = 0;
        for (int cnt = 1; cnt <= n; cnt += 2) {
            result += dp[n][cnt];
        }
        return result;
    }
    /**
     * @brief Compute the number of partitions of n into distinct odd parts (self-conjugate partitions)
     * @param n The integer to partition
     * @return The number of partitions into distinct odd parts
     */
    int countDistinctOddPartitions(int n) {
        // dp[i][j]: number of partitions of i into distinct odd parts, max part <= j
        std::vector<std::vector<int>> dp(n + 1, std::vector<int>(n + 1, 0));
        for (int j = 1; j <= n; j += 2) {
            if (j <= n) dp[j][j] = 1;
        }
        dp[0][0] = 1;
        for (int sum = 1; sum <= n; ++sum) {
            for (int maxPart = 1; maxPart <= n; maxPart += 2) {
                if (sum - maxPart >= 0) {
                    for (int prev = maxPart - 2; prev >= 0; prev -= 2) {
                        dp[sum][maxPart] += dp[sum - maxPart][prev];
                    }
                }
            }
        }
        int result = 0;
        for (int maxPart = 1; maxPart <= n; maxPart += 2) {
            result += dp[n][maxPart];
        }
        return result;
    }
} // namespace core

int main() {
    int n = 0;
    std::ifstream fin("input.txt");
    if (!fin) {
        std::cerr << "Error: Cannot open input.txt\n";
        return 1;
    }
    fin >> n;
    fin.close();
    if (n < 0) {
        std::cout << "n must be a non-negative integer.\n";
        return 0;
    }
    int oddDistinct = core::countOddDistinctPartitions(n);
    int selfConjugate = core::countDistinctOddPartitions(n);
    std::cout << "Number of partitions of " << n << " into an odd number of distinct parts: " << oddDistinct << std::endl;
    std::cout << "Number of self-conjugate partitions of " << n << " (partitions into distinct odd parts): " << selfConjugate << std::endl;
    if (oddDistinct == selfConjugate) {
        std::cout << "They are equal." << std::endl;
    } else {
        std::cout << "They are different." << std::endl;
    }
    return 0;
} 